### ISSUE TITLE
[Suggestion]  Display the name of type alias in quick-info consistently

### ISSUE THREAD

**Issue by HerringtonDarkholme** [Microsoft/TypeScript#13095]
I have tried my best to search existing issues. Pardon me if it is still duplicated.

Currently quick-info will always display the name of an `interface` no matter if it is imported from other modules or declared in the same file. This greatly shortens quick-info message length. 

Type alias behaves partly like `interface`: if type alias is declared in a non-module file, quick-info displays its alias name. 
```ts
type Alias<K> = {
  property1: K
} | {
  property2: number
}

var a: Alias<string> // popup as Alias<string>
```
[Online example](http://www.typescriptlang.org/play/#src=type%20Alias%3CK%3E%20%3D%20%7B%0A%20%20property1%3A%20K%0A%7D%20%7C%20%7B%0A%20%20property2%3A%20number%0A%7D%0A%0Avar%20a%3A%20Alias%3Cstring%3E)

However, an `import` statement will make type alias displayed as its type literal expansion. 

```ts
import 'test'
type Alias<K> = {
  property1: K
} | {
  property2: number
}

var a: Alias<string> // displayed as `{property1: string}|{property2: number}`
```
[example](http://www.typescriptlang.org/play/#src=import%20%27test%27%0Atype%20Alias%3CK%3E%20%3D%20%7B%0A%20%20property1%3A%20K%0A%7D%20%7C%20%7B%0A%20%20property2%3A%20number%0A%7D%0A%0Avar%20a%3A%20Alias%3Cstring%3E)

When used with mapped types, type alias' full expansion soon becomes very lengthy. For my extreme case, one type can have 7500+ lines of code after expansion. 

After mapped types have more adoption, displaying type alias, in my speculation, might be a real world usability issue. It might even strangulate language service if the type is too large to show. For example, https://github.com/Microsoft/TypeScript/issues/12904 is using mapped type for MongoDB API. While all the typing looks proper and valid to me, the quick-info for `MongoFilter<Person>` is quite daunting. Also, compiler error reports the alias name, not expansion.

It would be fine if TypeScript can show alias name in quick-info consistently and concisely.

**Comment by RyanCavanaugh** [comment]
This is a bit tricky due to implementation details.

Internally types are interned, and aliases are eagerly resolved, so a resolved definition of `Alias<K>` in `foo.ts` is indistinguishable from an identical declaration `Alias2<K>` in `bar.ts`.

When a type alias exists in the global scope, we store the first-seen alias with the type itself. This is a good heuristic (namely, aliases from lib.d.ts always "win" over user definitions of the same type). But we don't store the names of module-declared aliases because the same type alias might have different local names in different modules.

We could potentially keep a per-module type->alias lookup.

**Comment by RyanCavanaugh** [comment]
I was wrong about how the alias storage works - we don't care if it's global or not. Instead the problem is that the type-to-string code checks that the alias is reachable from the global scope, rather than the scope of the quickinfo request. It may be an issue of just using the correct reachability analysis depending on the QI invocation location. Investigate as Bug and see what happens

**Comment by kpdonn** [comment]
Heads up I submitted a PR that would fix this at #23638.

**Comment by RyanCavanaugh** [comment]
Nice to have but there is a lot of implementation complexity behind this. Still open to targeted PRs but I don't think we can make any specific promises here.
